### PR TITLE
[DataFrame] Implement .fillna(), .ffill(), .bfill(), .eval(), and .drop()

### PR DESCRIPTION
**Pull request by SaladRaider**
## What do these changes do?
Implements the .fillna(), .ffill(), .bfill(), .eval(), and .drop() methods for the ray df.

## Related issue number
Waiting on (https://issues.apache.org/jira/browse/ARROW-2122) to enable some tests involving dates and timestamps.

### CODE REVIEW

File: python/ray/dataframe/dataframe.py [hunk 1]
```diff
@@ -555,16 +621,188 @@ def expanding(self, min_periods=1, freq=None, center=False, axis=0):
...
+                    Returns:
+                        A pandas df with it's ith column replaced with
+                        the original column
+                    """
+                    n_df = ray.get(new_df._df[j]).copy()
+                    i_df = ray.get(identity._df[j]).copy()
+                    n_df[[n_df.columns[i]]] = i_df[[i_df.columns[i]]]
+                    return n_df
+
+                @ray.remote
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: @robertnishihara Should we allow inner functions to be defined as remote?

File: python/ray/dataframe/dataframe.py [hunk 2]
```diff
@@ -555,16 +621,188 @@ def expanding(self, min_periods=1, freq=None, center=False, axis=0):
...
+        new_df = self._map_partitions(
+            lambda df: df.fillna(value=value, method=method, axis=axis,
+                                 limit=limit, downcast=downcast, **kwargs)
+        )
+        is_bfill = method is not None and method in ['backfill', 'bfill']
+        is_ffill = method is not None and method in ['pad', 'ffill']
+        is_axis_zero = axis is None or axis == 0
+
+        if is_axis_zero:
+            if is_bfill or is_ffill:
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: I would prefer this not be nested like this. I think it would be more readable as:

```python
if is_axis_zero and (is_bfill or is_ffill):
    ...
else:
    ...
```

File: python/ray/dataframe/dataframe.py [hunk 3]
```diff
@@ -555,16 +621,188 @@ def expanding(self, min_periods=1, freq=None, center=False, axis=0):
...
+        if inplace:
+            self._df = new_df._df
+            self.columns = new_df.columns
+        else:
+            return new_df
 
     def fillna(self, value=None, method=None, axis=None, inplace=False,
                limit=None, downcast=None, **kwargs):
-        raise NotImplementedError("Not Yet implemented.")
+        if isinstance(value, (list, tuple)):
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: We need some docs here.

Also, this one is really long. Let's go over it together.

**Review by devin-petersohn — COMMENTED** [review]
Let's schedule a time to sit down together and go over this. I have a few questions.

File: python/ray/dataframe/dataframe.py (hunk 1 above)
Comment (robertnishihara) [reply python/ray/dataframe/dataframe.py]: Hmm... it should work.. but there is overhead every time a remote function is defined. So every time the `fillna` method is called, this will store the remote function definition in Redis and broadcast the definition to all of the workers. Whereas if it defined outside then that will only happen once.

How hard would it be to define this as a helper function somewhere outside of the method invocation?

Comment (SaladRaider) [reply python/ray/dataframe/dataframe.py]: It should not be too hard to do that. I define it outside.

File: python/ray/dataframe/dataframe.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
 from __future__ import print_function
 
 import pandas as pd
+
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Please remove extra newline

File: python/ray/dataframe/dataframe.py [hunk 5]
```diff
@@ -17,6 +18,7 @@
 import warnings
 import numpy as np
 import ray
+
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Please remove extra newline

File: python/ray/dataframe/dataframe.py [hunk 6]
```diff
@@ -36,6 +38,7 @@ def __init__(self, df, columns, index=None):
 
         self._df = df
         self.columns = columns
+        self._pd_index = None
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Please remove extra code

File: python/ray/dataframe/dataframe.py [hunk 7]
```diff
@@ -130,6 +133,23 @@ def _set_lengths(self, lengths):
 
     _lengths = property(_get_lengths, _set_lengths)
 
+    def _map_to_partition_range_index(self, pd_index, partition_idx):
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: is this needed with the `loc` way we are doing things now?

File: python/ray/dataframe/dataframe.py [hunk 8]
```diff
@@ -222,6 +242,7 @@ def _map_partitions(self, func, index=None):
         """
         assert(callable(func))
         new_df = [_deploy_func.remote(func, part) for part in self._df]
+
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Please remove extra newline

File: python/ray/dataframe/dataframe.py [hunk 9]
```diff
@@ -458,8 +479,6 @@ def dropna(self, axis, how, thresh=None, subset=[], inplace=False):
             DataFrame with the dropna applied.
         """
         raise NotImplementedError("Not yet")
-        if how != 'any' and how != 'all':
```
Comment (devin-petersohn) [inline python/ray/dataframe/dataframe.py]: Why was this deleted?

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
It's looking good! We're really close.

File: python/ray/dataframe/dataframe.py (hunk 9 above)
Comment (SaladRaider) [reply python/ray/dataframe/dataframe.py]: I removed it because it will never be reached because of the `NotImplementedError`

**Comment by devin-petersohn** [comment]
Test failure is unrelated. Thanks @SaladRaider!

**Review by devin-petersohn — APPROVED** [review]
